### PR TITLE
SAMZA-1219 Add metrics for operator message received and execution times

### PR DESCRIPTION
**Pull request by prateekm**

### CODE REVIEW

**Comment by prateekm** [comment]
@vjagadish1989 Please take a look.

File: samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java [hunk 1]
```diff
@@ -29,60 +37,136 @@
...
+   * Delegates to {@link #handleTimer(MessageCollector, TaskCoordinator)} for handling the timer tick.
    *
    * @param collector  the {@link MessageCollector} in the context
    * @param coordinator  the {@link TaskCoordinator} in the context
    */
-  public final void onTick(MessageCollector collector, TaskCoordinator coordinator) {
-    onTimer(collector, coordinator);
-    nextOperators.forEach(sub -> sub.onTick(collector, coordinator));
+  public final void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
+    long startNs = System.nanoTime();
```
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java]: IIUC, We have a mode in `HDFS` to disable timers entirely. That way, we don't make a lot of calls to `System.nanoTime()`. 

Would be nice to support that as a part of a different PR.

File: samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java [hunk 2]
```diff
@@ -29,60 +37,136 @@
...
+   * Get the {@link OperatorSpec} for this {@link OperatorImpl}.
    *
-   * @param outputMessage  output message
-   * @param collector  the {@link MessageCollector} in the context
-   * @param coordinator  the {@link TaskCoordinator} in the context
+   * @return the {@link OperatorSpec} for this {@link OperatorImpl}
    */
-  void propagateResult(RM outputMessage, MessageCollector collector, TaskCoordinator coordinator) {
-    nextOperators.forEach(sub -> sub.onNext(outputMessage, collector, coordinator));
+  protected abstract OperatorSpec<RM> getOpSpec();
```
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java]: nit: `s/getOpSpec/getOperatorSpec`?

File: samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java [hunk 3]
```diff
@@ -163,24 +164,22 @@ public RootOperatorImpl getRootOperator(SystemStream systemStream) {
...
+   * @param <M>  type of input message
    * @return  the {@link OperatorImpl} implementation instance
    */
-  private <M> OperatorImpl<M, ?> createOperatorImpl(MessageStreamImpl<M> source,
-      OperatorSpec operatorSpec, Config config, TaskContext context) {
+  private <M> OperatorImpl<M, ?> createOperatorImpl(OperatorSpec operatorSpec, Config config, TaskContext context) {
     if (operatorSpec instanceof StreamOperatorSpec) {
       StreamOperatorSpec<M, ?> streamOpSpec = (StreamOperatorSpec<M, ?>) operatorSpec;
-      return new StreamOperatorImpl<>(streamOpSpec, source, config, context);
+      return new StreamOperatorImpl<>(streamOpSpec, config, context);
```
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java]: Very minor: Can potentially inline as the rest?

File: samza-core/src/main/java/org/apache/samza/operators/impl/PartialJoinOperatorImpl.java [hunk 4]
```diff
@@ -47,35 +50,41 @@
...
+    this.thisPartialJoinFn.init(config, context);
+  }
+
+  @Override
+  public Collection<RM> handleMessage(M message, MessageCollector collector, TaskCoordinator coordinator) {
     K key = thisPartialJoinFn.getKey(message);
     thisPartialJoinFn.getState().put(key, new PartialJoinMessage<>(message, clock.currentTimeMillis()));
     PartialJoinMessage<JM> otherMessage = otherPartialJoinFn.getState().get(key);
     long now = clock.currentTimeMillis();
     if (otherMessage != null && otherMessage.getReceivedTimeMs() > now - ttlMs) {
```
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/impl/PartialJoinOperatorImpl.java]: Is it useful to add a `matchedKeys` metric for the join operator? Seems like a useful one to track when doing joins to get some visibility into join correctness.

File: samza-core/src/main/java/org/apache/samza/operators/impl/PartialJoinOperatorImpl.java [hunk 5]
```diff
@@ -47,35 +50,41 @@
...
+    this.partialJoinOpSpec = partialJoinOpSpec;
+    this.thisPartialJoinFn = partialJoinOpSpec.getThisPartialJoinFn();
+    this.otherPartialJoinFn = partialJoinOpSpec.getOtherPartialJoinFn();
+    this.ttlMs = partialJoinOpSpec.getTtlMs();
     this.clock = clock;
   }
 
   @Override
-  public void onNext(M message, MessageCollector collector, TaskCoordinator coordinator) {
+  protected void doInit(Config config, TaskContext context) {
```
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/impl/PartialJoinOperatorImpl.java]: Very minor:
`handleInit` for consistency with `handleMessage` and `handleTimer`?

File: samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java [hunk 6]
```diff
@@ -74,22 +78,35 @@
...
 
+  private final WindowOperatorSpec<M, WK, WV> windowOpSpec;
+  private final Clock clock;
   private final WindowInternal<M, WK, WV> window;
   private final KeyValueStore<WindowKey<WK>, WindowState<WV>> store = new InternalInMemoryStore<>();
-  TriggerScheduler<WK> triggerScheduler ;
+  private TriggerScheduler<WK> triggerScheduler;
+
+  // Results to be returned for the current handleMessage and handleTimer call.
+  private List<WindowPane<WK, WV>> currentResults = new ArrayList<>();
```
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java]: Wondering if it'll be slightly cleaner to remove this, and have individual methods return a `List<WindowPane>` objects that'd fire? 

Perhaps, it's not super-obvious to me, Is there a need to book-keep the current `handleMessage` results? It seems that we are emitting them lazily anyways?  (as opposed to the current implementation that propagates results down the chain eagerly, and mutates the state store later)

File: samza-core/src/main/java/org/apache/samza/operators/spec/PartialJoinOperatorSpec.java [hunk 7]
```diff
@@ -88,10 +86,4 @@ public long getTtlMs() {
   public int getOpId() {
     return this.opId;
   }
-
-  @Override
```
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/spec/PartialJoinOperatorSpec.java]: thanks!

File: samza-core/src/test/java/org/apache/samza/operators/impl/TestOperatorImpl.java [hunk 8]
```diff
@@ -18,61 +18,205 @@
...
-    assertEquals(mockCoordinator, this.curCoordinator);
+    opImpl.onTimer(mockCollector, mockCoordinator);
+
+    // verify that it updates metrics
+    verify(mockMessageCounter, times(0)).inc();
+    verify(mockTimer, times(1)).update(anyLong());
+  }
+
+  private static class TestOpImpl extends OperatorImpl<Object, Object> {
+    private final Object mockOutput;
```
Comment (vjagadish1989) [inline samza-core/src/test/java/org/apache/samza/operators/impl/TestOperatorImpl.java]: AFAICT, `mockOutput` appears to be unused? Or do some tests assert the behavior that `mockOutput` is actually contained in the emitted result?

File: samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java [hunk 9]
```diff
@@ -29,60 +37,136 @@
  * Abstract base class for all stream operator implementations.
  */
 public abstract class OperatorImpl<M, RM> {
+  private static final String METRICS_GROUP = OperatorImpl.class.getName();
 
-  private final Set<OperatorImpl<RM, ?>> nextOperators = new HashSet<>();
+  private Set<OperatorImpl<RM, ?>> registeredOperators;
```
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java]: Sure, sounds good. thanks!

File: samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java (hunk 1 above)
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java]: Thanks!

File: samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java [hunk 10]
```diff
@@ -29,60 +37,136 @@
...
   /**
-   * Invoked at every tick. Implementations must call {@link #propagateResult} to propagate any generated output
-   * to registered downstream operators.
+   * Handle the the timer tick for this operator and return the results to be propagated to registered operators.
+   * <p>
+   * Defaults to a no-op implementation.
    *
    * @param collector  the {@link MessageCollector} in the context
    * @param coordinator  the {@link TaskCoordinator} in the context
+   * @return  results of the timed operation
```
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java]: Ah, this is my moment of zen. I did not know that `Collections.emptyList` returns a cached, immutable list.

File: samza-core/src/main/java/org/apache/samza/operators/impl/PartialJoinOperatorImpl.java (hunk 4 above)
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/impl/PartialJoinOperatorImpl.java]: Yeah, makes sense. Thanks for pointing out! +1 for adding `keysRemoved` metric.  

It does seem quite useful for in-memory stores / TTL implemented at the app layer (not entirely sure on the portability across TTL implementations and stores).

File: samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java (hunk 6 above)
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java]: I was thinking that the version that returns `List<WindowPane>` might be slightly cleaner. But, now that I see the implementation, It's not vastly different. +0 on this one. Your pick 👍

**Review by vjagadish1989 — APPROVED** [review]
approved, looks good to me. thanks for the contribution!

**Pull request closed without merging**